embassy-stm32/src/rng.rs

```diff
@@ -1,52 +1,170 @@
 #![macro_use]
-use crate::pac::rng::{regs, Rng};
+
+//use crate::pac::rng::{regs, Rng};
+use crate::pac;
 use crate::peripherals;
-use embassy::util::Unborrow;
+use crate::interrupt;
+use embassy::util::{Unborrow, AtomicWaker};
 use embassy_extras::unborrow;
+use rand_core::{RngCore, CryptoRng};
+
+use defmt::*;
+
+static RNG_WAKER: AtomicWaker = AtomicWaker::new();
+
+#[interrupt]
+unsafe fn RNG() {
+    let bits = crate::pac::RNG.sr().read();
+    if bits.drdy() || bits.seis() || bits.ceis() {
+        crate::pac::RNG.cr().write(|reg| reg.set_ie(false));
+        RNG_WAKER.wake();
+    }
+}
 
 pub struct Random<T: Instance> {
     inner: T,
 }
 
 impl<T: Instance> Random<T> {
-    pub fn new(inner: impl Unborrow<Target = T>) -> Self {
+    pub fn new(inner: impl Unborrow<Target=T>) -> Self {
         unborrow!(inner);
         Self { inner }
     }
 }
 
+impl<T: Instance> RngCore for Random<T> {
+    fn next_u32(&mut self) -> u32 {
+        loop {
+            let bits = unsafe { T::regs().sr().read() };
+            if bits.drdy() {
+                return unsafe{ T::regs().dr().read() }
+            }
+        }
+    }
+
+    fn next_u64(&mut self) -> u64 {
+        let mut rand = self.next_u32() as u64;
+        rand |= (self.next_u32() as u64) << 32;
+        rand
+    }
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        for chunk in dest.chunks_mut(4) {
+            let rand = self.next_u32();
+            for (slot, num) in chunk.iter_mut().zip(rand.to_be_bytes().iter()) {
+                *slot = *num
+            }
+        }
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand_core::Error> {
+        self.fill_bytes( dest );
+        Ok(())
+    }
+}
+
+impl<T: Instance> CryptoRng for Random<T> { }
+
 use core::future::Future;
 use core::marker::PhantomData;
-use embassy::traits::rng::Rng as RngTrait;
+use embassy::traits;
+use core::task::{Poll, Context};
+use core::pin::Pin;
+
+struct RngInterruptFuture<T: Instance> {
```

Comment:
It's nicer to use `futures::poll_fn`, inline in the `fill_bytes` code. It reduces the boilerplate of creating a future manually, and it makes the code easier to read, since it reads top-to-bottom. See [this](https://github.com/embassy-rs/embassy/blob/76f737bb0e62146cecba618a34e9efc8c91fb3c3/embassy-nrf/src/uarte.rs#L266-L273) for an example